### DISCUSSION TITLE
User defined nonces and transaction dependencies

### DISCUSSION THREAD

**Discussion by auryn-macmillan** [5afe/safe-core-protocol-specs#45] [Ideas]
Safe nonces for replay protection. Currently, in 1.x, nonces increment after each successful transaction, allowing some predictability on the order in which transactions are executed. But also meaning users must predict the safe nonce at the time of execution when signing.

Managing the queue of safe transactions is often a pain point when, for one reason or another, the predicted nonce differs from the actual nonce.

What if, rather than using a nonce that increments on each successful transaction, we let users define a nonce, along with an array of the hashes of any previous transactions which must have successfully executed before this transaction can be executed. This way users could define sequences of transactions, but would not need to manage a fully sequential queue of transactions and would no longer need to replace transactions that will always revert.

One consideration is that if an approved transaction is executable, but no longer wanted, it would need to be marked as invalid.

**Comment by mmv08** [comment]
The protocol doesn't manage replay protection, so it should be possible to implement such a module that would take an arbitrary nonce
